x/programs/rust/examples/nft/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+        .expect("failed to store total supply");
+
+    true
+}
+
+/// Mints NFT tokens and sends them to the recipient.
+#[public]
+pub fn mint(program: Program, recipient: Address, quantity: i64) -> bool {
+    assert!(quantity > 0, "quantity must be greater than zero");
+    assert!(
+        quantity <= NFT_TOTAL_SUPPLY as i64,
+        "quantity must be less than or equal to total supply"
+    );
+
+    let balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(recipient).to_vec())
+        .expect("failed to get balance");
+
+    program
+        .state()
+        .store(StateKey::Balance(recipient).to_vec(), &(balance + quantity))
+        .expect("failed to store balance");
+
+    true
+}
+
+/// Transfers balance from the sender to the the recipient.
+#[public]
+pub fn transfer(program: Program, sender: Address, recipient: Address, amount: i64) -> bool {
+    assert_eq!(sender, recipient, "sender and recipient must be different");
+
+    // ensure the sender has adequate balance
+    let sender_balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(sender).to_vec())
+        .expect("failed to update balance");
+
+    assert!(
+        amount >= 0 && sender_balance >= amount,
+        "sender and recipient must be different"
+    );
+
+    assert_eq!(sender, recipient, "sender and recipient must be different");
+
+    let recipient_balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(recipient).to_vec())
+        .expect("failed to store balance");
+
+    // update balances
+    program
+        .state()
+        .store(
+            StateKey::Balance(sender).to_vec(),
+            &(sender_balance - amount),
+        )
+        .expect("failed to store balance");
+
+    program
+        .state()
+        .store(
+            StateKey::Balance(recipient).to_vec(),
+            &(recipient_balance + amount),
+        )
+        .expect("failed to store balance");
+
+    true
+}
+
+#[public]
+pub fn burn(program: Program, from: Address, amount_burned: i64) -> bool {
+    assert!(amount_burned > 0, "quantity must be greater than zero");
+
+    let balance = program
+        .state()
+        .get::<i64, _>(StateKey::Balance(from).to_vec())
+        .expect("failed to get balance");
+
+    assert!(
+        amount_burned <= balance,
+        "quantity must be less than or equal to total supply"
+    );
+
+    program
+        .state()
+        .store(StateKey::Balance(from).to_vec(), &(balance - amount_burned))
+        .expect("failed to store new balance");
+
+    true
+}
+
+/// The program state keys.
+#[state_keys]
+enum StateKey {
+    /// The total supply of the token. Key prefix 0x0.
+    TotalSupply,
+    /// The name of the token. Key prefix 0x1.
+    Name,
+    /// The symbol of the token. Key prefix 0x2.
+    Symbol,
+    /// Metadata of the token. Key prefix 0x3.
+    Metadata,
+    /// Owner address. Key prefix 0x4(address).
+    Owner(Address),
```

Comment:
I **think** the program will have a single owner then each `NFT(Address)` = metadata_bytes?

Reply:
If that is true I think we might need another step here that creates a bunch of NFTs 1 tx per. Then mint_to saves the metadata? Needs some thought.

Reply:
This is a singleton NFT contract. Only one possible owner of the NFT at a time.